Cloning call to method via clad::utils::StmtClone::VisitCXXMemberCallExpr causes segmentation fault

Can be reproduced by
```
struct S{
  int f0() { return 0; }
  int f1(int x) { return f0() + x; }
};
...
clad::differentiate(&S::f1, 0);
```

It no longer causes segfault on master, now just a warning
`issue49.cpp:5:26: warning: function 'f0' was not differentiated because clad failed to differentiate it and no overload was found in namespace 'custom_derivatives'`

But it's reproducible with such revisions
```
clang: ae610a215a867aadf97fc8cf9a61f9ffe4901476
llvm: release_50
clad: 11777b8
```

<details>
<summary>output</summary>

```
➜  demos git:(2b6298e) ✗ /home/blide/gsoc/inst/bin/clang-5.0 -Xclang -add-plugin -Xclang clad -Xclang -load -Xclang /home/blide/gsoc/inst/lib/clad.so -I../include/ -x c++ -std=c++11 issue49.cpp
#0 0x000055d55b884ad7 llvm::sys::PrintStackTrace(llvm::raw_ostream&) /home/blide/gsoc/src/lib/Support/Unix/Signals.inc:398:22
#1 0x000055d55b884b6a PrintStackTraceSignalHandler(void*) /home/blide/gsoc/src/lib/Support/Unix/Signals.inc:462:1
#2 0x000055d55b882d97 llvm::sys::RunSignalHandlers() /home/blide/gsoc/src/lib/Support/Signals.cpp:49:19
#3 0x000055d55b884345 SignalHandler(int) /home/blide/gsoc/src/lib/Support/Unix/Signals.inc:252:1
#4 0x00007f5411643f40 __restore_rt (/lib/x86_64-linux-gnu/libpthread.so.0+0x13f40)
#5 0x000055d55bb94286 clang::Expr::isTypeDependent() const /home/blide/gsoc/src/tools/clang/include/clang/AST/Expr.h:166:50
#6 0x000055d55ea42ac2 clang::CallExpr::updateDependenciesFromArg(clang::Expr*) /home/blide/gsoc/src/tools/clang/lib/AST/Expr.cpp:1206:3
#7 0x000055d55ea427e2 clang::CallExpr::CallExpr(clang::ASTContext const&, clang::Stmt::StmtClass, clang::Expr*, llvm::ArrayRef<clang::Expr*>, llvm::ArrayRef<clang::Expr*>, clang::QualType, clang::ExprValueKind, clang::SourceLocation) /home/blide/gsoc/src/tools/clang/lib/AST/Expr.cpp:1177:50
#8 0x000055d55ea428d9 clang::CallExpr::CallExpr(clang::ASTContext const&, clang::Stmt::StmtClass, clang::Expr*, llvm::ArrayRef<clang::Expr*>, clang::QualType, clang::ExprValueKind, clang::SourceLocation) /home/blide/gsoc/src/tools/clang/lib/AST/Expr.cpp:1187:69
#9 0x00007f5410fc8aa8 clang::CXXMemberCallExpr::CXXMemberCallExpr(clang::ASTContext&, clang::Expr*, llvm::ArrayRef<clang::Expr*>, clang::QualType, clang::ExprValueKind, clang::SourceLocation) /home/blide/gsoc/src/tools/clad/../clang/include/clang/AST/ExprCXX.h:140:62
#10 0x00007f5410fc4634 clad::utils::StmtClone::VisitCXXMemberCallExpr(clang::CXXMemberCallExpr*) /home/blide/gsoc/src/tools/clad/lib/Differentiator/StmtClone.cpp:205:77
#11 0x00007f5410ef16a0 clang::StmtVisitorBase<clang::make_ptr, clad::utils::StmtClone, clang::Stmt*>::Visit(clang::Stmt*) /home/blide/gsoc/obj/tools/clad/../clang/include/clang/AST/StmtNodes.inc:339:1
#12 0x00007f5410fc9922 clang::Expr* clad::utils::StmtClone::Clone<clang::Expr>(clang::Expr const*) /home/blide/gsoc/src/tools/clad/include/clad/Differentiator/StmtClone.h:130:18
#13 0x00007f5410fc0fd0 clad::utils::StmtClone::VisitBinaryOperator(clang::BinaryOperator*) /home/blide/gsoc/src/tools/clad/lib/Differentiator/StmtClone.cpp:44:1
#14 0x00007f5410efcebd clang::StmtVisitorBase<clang::make_ptr, clad::utils::StmtClone, clang::Stmt*>::VisitBinAdd(clang::BinaryOperator*) /home/blide/gsoc/src/tools/clad/../clang/include/clang/AST/StmtVisitor.h:125:3
#15 0x00007f5410ef0d7f clang::StmtVisitorBase<clang::make_ptr, clad::utils::StmtClone, clang::Stmt*>::Visit(clang::Stmt*) /home/blide/gsoc/src/tools/clad/../clang/include/clang/AST/StmtVisitor.h:52:26
#16 0x00007f5410ee38d3 clang::Stmt* clad::utils::StmtClone::Clone<clang::Stmt>(clang::Stmt const*) /home/blide/gsoc/src/tools/clad/include/clad/Differentiator/StmtClone.h:130:18
#17 0x00007f5410ec4210 clad::DerivativeBuilder::Clone(clang::Stmt const*) /home/blide/gsoc/src/tools/clad/lib/Differentiator/DerivativeBuilder.cpp:220:43
#18 0x00007f5410ec4374 clad::ForwardModeVisitor::Clone(clang::Stmt const*) /home/blide/gsoc/src/tools/clad/lib/Differentiator/DerivativeBuilder.cpp:238:42
#19 0x00007f5410ec4980 clad::ForwardModeVisitor::VisitReturnStmt(clang::ReturnStmt const*) /home/blide/gsoc/src/tools/clad/lib/Differentiator/DerivativeBuilder.cpp:288:58
#20 0x00007f5410ee379b clang::StmtVisitorBase<clang::make_const_ptr, clad::ForwardModeVisitor, clad::NodeContext>::Visit(clang::Stmt const*) /home/blide/gsoc/obj/tools/clad/../clang/include/clang/AST/StmtNodes.inc:1219:1
#21 0x00007f5410ec447e clad::ForwardModeVisitor::VisitCompoundStmt(clang::CompoundStmt const*) /home/blide/gsoc/src/tools/clad/lib/Differentiator/DerivativeBuilder.cpp:249:40
#22 0x00007f5410ee242d clang::StmtVisitorBase<clang::make_const_ptr, clad::ForwardModeVisitor, clad::NodeContext>::Visit(clang::Stmt const*) /home/blide/gsoc/obj/tools/clad/../clang/include/clang/AST/StmtNodes.inc:79:1
#23 0x00007f5410ec40a2 clad::ForwardModeVisitor::Derive(clad::FunctionDeclInfo&, clad::DiffPlan const&) /home/blide/gsoc/src/tools/clad/lib/Differentiator/DerivativeBuilder.cpp:203:77
#24 0x00007f5410ec351f clad::DerivativeBuilder::Derive(clad::FunctionDeclInfo&, clad::DiffPlan const&) /home/blide/gsoc/src/tools/clad/lib/Differentiator/DerivativeBuilder.cpp:73:24
#25 0x00007f5410ebe7d8 clad::plugin::CladPlugin::HandleTopLevelDecl(clang::DeclGroupRef) /home/blide/gsoc/src/tools/clad/tools/ClangPlugin.cpp:133:51
#26 0x000055d55c1b1466 clang::MultiplexConsumer::HandleTopLevelDecl(clang::DeclGroupRef) /home/blide/gsoc/src/tools/clang/lib/Frontend/MultiplexConsumer.cpp:284:25
#27 0x000055d55d432a4b clang::ParseAST(clang::Sema&, bool, bool) /home/blide/gsoc/src/tools/clang/lib/Parse/ParseAST.cpp:151:18
#28 0x000055d55c17e6ff clang::ASTFrontendAction::ExecuteAction() /home/blide/gsoc/src/tools/clang/lib/Frontend/FrontendAction.cpp:1003:11
#29 0x000055d55c67730e clang::CodeGenAction::ExecuteAction() /home/blide/gsoc/src/tools/clang/lib/CodeGen/CodeGenAction.cpp:993:1
#30 0x000055d55c17e142 clang::FrontendAction::Execute() /home/blide/gsoc/src/tools/clang/lib/Frontend/FrontendAction.cpp:906:38
#31 0x000055d55c11d898 clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) /home/blide/gsoc/src/tools/clang/lib/Frontend/CompilerInstance.cpp:981:24
#32 0x000055d55c2c472c clang::ExecuteCompilerInvocation(clang::CompilerInstance*) /home/blide/gsoc/src/tools/clang/lib/FrontendTool/ExecuteCompilerInvocation.cpp:251:38
#33 0x000055d55a63b9a9 cc1_main(llvm::ArrayRef<char const*>, char const*, void*) /home/blide/gsoc/src/tools/clang/tools/driver/cc1_main.cpp:221:38
#34 0x000055d55a630ee2 ExecuteCC1Tool(llvm::ArrayRef<char const*>, llvm::StringRef) /home/blide/gsoc/src/tools/clang/tools/driver/driver.cpp:306:64
#35 0x000055d55a631a78 main /home/blide/gsoc/src/tools/clang/tools/driver/driver.cpp:387:26
#36 0x00007f54110cfb6b __libc_start_main /build/glibc-KRRWSm/glibc-2.29/csu/../csu/libc-start.c:308:16
#37 0x000055d55a62e61a _start (/home/blide/gsoc/inst/bin/clang-5.0+0x163e61a)
Stack dump:
0.	Program arguments: /home/blide/gsoc/inst/bin/clang-5.0 -cc1 -triple x86_64-unknown-linux-gnu -emit-obj -mrelax-all -disable-free -main-file-name issue49.cpp -mrelocation-model static -mthread-model posix -mdisable-fp-elim -fmath-errno -masm-verbose -mconstructor-aliases -munwind-tables -fuse-init-array -target-cpu x86-64 -dwarf-column-info -debugger-tuning=gdb -resource-dir /home/blide/gsoc/inst/lib/clang/5.0.2 -I ../include/ -internal-isystem /usr/lib/gcc/x86_64-linux-gnu/8/../../../../include/c++/8 -internal-isystem /usr/lib/gcc/x86_64-linux-gnu/8/../../../../include/x86_64-linux-gnu/c++/8 -internal-isystem /usr/lib/gcc/x86_64-linux-gnu/8/../../../../include/x86_64-linux-gnu/c++/8 -internal-isystem /usr/lib/gcc/x86_64-linux-gnu/8/../../../../include/c++/8/backward -internal-isystem /usr/local/include -internal-isystem /home/blide/gsoc/inst/lib/clang/5.0.2/include -internal-externc-isystem /usr/include/x86_64-linux-gnu -internal-externc-isystem /include -internal-externc-isystem /usr/include -std=c++11 -fdeprecated-macro -fdebug-compilation-dir /home/blide/gsoc/src/tools/clad/demos -ferror-limit 19 -fmessage-length 118 -fobjc-runtime=gcc -fcxx-exceptions -fexceptions -fdiagnostics-show-option -fcolor-diagnostics -add-plugin clad -load /home/blide/gsoc/inst/lib/clad.so -o /tmp/issue49-e7aad9.o -x c++ issue49.cpp 
1.	<eof> parser at end of file
clang-5.0: error: unable to execute command: Segmentation fault (core dumped)
clang-5.0: error: clang frontend command failed due to signal (use -v to see invocation)
clang version 5.0.2 (https://github.com/llvm-mirror/clang.git ae610a215a867aadf97fc8cf9a61f9ffe4901476) (https://github.com/llvm-mirror/llvm.git da4a2839d80ac52958be0129b871beedfe90136e)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /home/blide/gsoc/inst/bin
clang-5.0: note: diagnostic msg: PLEASE submit a bug report to http://llvm.org/bugs/ and include the crash backtrace, preprocessed source, and associated run script.
clang-5.0: note: diagnostic msg: 
********************

PLEASE ATTACH THE FOLLOWING FILES TO THE BUG REPORT:
Preprocessed source(s) and associated run script(s) are located at:
clang-5.0: note: diagnostic msg: /tmp/issue49-f9238a.cpp
clang-5.0: note: diagnostic msg: /tmp/issue49-f9238a.sh
clang-5.0: note: diagnostic msg: 

********************
```
</details>

--- a/test/FirstDerivative/StructMethodCall.C
+++ b/test/FirstDerivative/StructMethodCall.C
@@ -1,4 +1,5 @@
-// RUN: %cladclang %s -I%S/../../include 2>&1 | FileCheck %s
+// RUN: %cladclang %s -I%S/../../include -oStructMethodCall.out 2>&1 | FileCheck %s
+// RUN: ./StructMethodCall.out | FileCheck -check-prefix=CHECK-EXEC %s
 //CHECK-NOT: {{.*error|warning|note:.*}}
 
 #include "clad/Differentiator/Differentiator.h"
@@ -12,10 +13,10 @@ public:
   }
 
   // CHECK: int f_darg0(int x) {
-  // CHECK-NEXT: int _d_x = 1;
+  // CHECK-NEXT:     int _d_x = 1;
   // CHECK-NEXT:     A _d_this_obj;
   // CHECK-NEXT:     A *_d_this = &_d_this_obj;
-  // CHECK-NEXT: return _d_x;
+  // CHECK-NEXT:     return _d_x;
   // CHECK-NEXT: }
 
   int g_1(int x, int y) {
@@ -23,19 +24,19 @@ public:
   }
 
   // CHECK: int g_1_darg0(int x, int y) {
-  // CHECK-NEXT: int _d_x = 1;
-  // CHECK-NEXT: int _d_y = 0;
+  // CHECK-NEXT:     int _d_x = 1;
+  // CHECK-NEXT:     int _d_y = 0;
   // CHECK-NEXT:     A _d_this_obj;
   // CHECK-NEXT:     A *_d_this = &_d_this_obj;
-  // CHECK-NEXT: return _d_x * x + x * _d_x + _d_y;
+  // CHECK-NEXT:     return _d_x * x + x * _d_x + _d_y;
   // CHECK-NEXT: }
 
   // CHECK: int g_1_darg1(int x, int y) {
-  // CHECK-NEXT: int _d_x = 0;
-  // CHECK-NEXT: int _d_y = 1;
+  // CHECK-NEXT:     int _d_x = 0;
+  // CHECK-NEXT:     int _d_y = 1;
   // CHECK-NEXT:     A _d_this_obj;
   // CHECK-NEXT:     A *_d_this = &_d_this_obj;
-  // CHECK-NEXT: return _d_x * x + x * _d_x + _d_y;
+  // CHECK-NEXT:     return _d_x * x + x * _d_x + _d_y;
   // CHECK-NEXT: }
 
   int g_2(int x, int y) {
@@ -43,35 +44,70 @@ public:
   }
 
   // CHECK: int g_2_darg0(int x, int y) {
-  // CHECK-NEXT: int _d_x = 1;
-  // CHECK-NEXT: int _d_y = 0;
+  // CHECK-NEXT:     int _d_x = 1;
+  // CHECK-NEXT:     int _d_y = 0;
   // CHECK-NEXT:     A _d_this_obj;
   // CHECK-NEXT:     A *_d_this = &_d_this_obj;
-  // CHECK-NEXT: return _d_x + _d_y * y + y * _d_y;
+  // CHECK-NEXT:     return _d_x + _d_y * y + y * _d_y;
   // CHECK-NEXT: }
 
   // CHECK: int g_2_darg1(int x, int y) {
-  // CHECK-NEXT: int _d_x = 0;
-  // CHECK-NEXT: int _d_y = 1;
+  // CHECK-NEXT:     int _d_x = 0;
+  // CHECK-NEXT:     int _d_y = 1;
   // CHECK-NEXT:     A _d_this_obj;
   // CHECK-NEXT:     A *_d_this = &_d_this_obj;
-  // CHECK-NEXT: return _d_x + _d_y * y + y * _d_y;
+  // CHECK-NEXT:     return _d_x + _d_y * y + y * _d_y;
   // CHECK-NEXT: }
 
 
   int m(int x, int y) {
     return f(x) + g_1(x, y);
   }
+
+  // CHECK: clad::ValueAndPushforward<int, int> f_pushforward(int x, A *_d_this, int _d_x) {
+  // CHECK-NEXT:     return {x, _d_x};
+  // CHECK-NEXT: }
+
+  // CHECK: clad::ValueAndPushforward<int, int> g_1_pushforward(int x, int y, A *_d_this, int _d_x, int _d_y) {
+  // CHECK-NEXT:     return {x * x + y, _d_x * x + x * _d_x + _d_y};
+  // CHECK-NEXT: }
+
+  // CHECK: int m_darg0(int x, int y) {
+  // CHECK-NEXT:     int _d_x = 1;
+  // CHECK-NEXT:     int _d_y = 0;
+  // CHECK-NEXT:     A _d_this_obj;
+  // CHECK-NEXT:     A *_d_this = &_d_this_obj;
+  // CHECK-NEXT:     clad::ValueAndPushforward<int, int> _t0 = this->f_pushforward(x, _d_this, _d_x);
+  // CHECK-NEXT:     clad::ValueAndPushforward<int, int> _t1 = this->g_1_pushforward(x, y, _d_this, _d_x, _d_y);
+  // CHECK-NEXT:     return _t0.pushforward + _t1.pushforward;
+  // CHECK-NEXT: }
+
+  // CHECK: int m_darg1(int x, int y) {
+  // CHECK-NEXT:     int _d_x = 0;
+  // CHECK-NEXT:     int _d_y = 1;
+  // CHECK-NEXT:     A _d_this_obj;
+  // CHECK-NEXT:     A *_d_this = &_d_this_obj;
+  // CHECK-NEXT:     clad::ValueAndPushforward<int, int> _t0 = this->f_pushforward(x, _d_this, _d_x);
+  // CHECK-NEXT:     clad::ValueAndPushforward<int, int> _t1 = this->g_1_pushforward(x, y, _d_this, _d_x, _d_y);
+  // CHECK-NEXT:     return _t0.pushforward + _t1.pushforward;
+  // CHECK-NEXT: }
 };
 
 int main () {
   A a;
-  clad::differentiate(&A::f, 0);
-  clad::differentiate(&A::g_1, 0);
-  clad::differentiate(&A::g_1, 1);
-  clad::differentiate(&A::g_2, 0);
-  clad::differentiate(&A::g_2, 1);
-  //clad::differentiate(&A::m, 0);
-  //clad::differentiate(&A::m, 1);
+  auto f_dx = clad::differentiate(&A::f, 0);
+  printf("Result is = %d\n", f_dx.execute(a, 1)); // CHECK-EXEC: Result is = 1
+
+  auto g_1_dx = clad::differentiate(&A::g_1, 0);
+  auto g_1_dy = clad::differentiate(&A::g_1, 1);
+  printf("Result is = {%d, %d}\n", g_1_dx.execute(a, 1, 2), g_1_dy.execute(a, 1, 2)); // CHECK-EXEC: Result is = {2, 1}
+
+  auto g_2_dx = clad::differentiate(&A::g_2, 0);
+  auto g_2_dy = clad::differentiate(&A::g_2, 1);
+  printf("Result is = {%d, %d}\n", g_2_dx.execute(a, 1, 2), g_2_dy.execute(a, 1, 2)); // CHECK-EXEC: Result is = {1, 4}
+
+  auto m_dx = clad::differentiate(&A::m, 0);
+  auto m_dy = clad::differentiate(&A::m, 1);
+  printf("Result is = {%d, %d}\n", m_dx.execute(a, 1, 2), m_dy.execute(a, 1, 2)); // CHECK-EXEC: Result is = {3, 1}
   return 0;
 }

